### ISSUE TITLE
Authorization header missing in Zuul proxy using Spring boot 1.3.5.RELEASE and Brixton.RC1/RC2

### ISSUE THREAD

**Issue by maazkhan** [spring-cloud/spring-cloud-netflix#1034]
I am creating a simple reverse proxy with OAuth support Using @EnableOAuth2Sso
@EnableZuulProxy following this example https://github.com/spring-guides/tut-spring-security-and-angular-js/blob/master/oauth2/ui/src/main/java/demo/UiApplication.java

I have Zuul "Pre filter" that simply prints headers. I noticed that the "authorization : bearer .." header is missing when processing the Zuul filter.

As per the docs this should be a case of simple client token relay http://projects.spring.io/spring-cloud/spring-cloud.html#_token_relay

Here is Zuul filter - https://gist.github.com/maazkhan/1e80e98610ccd46407456233bfc4d770
build.gradle - https://gist.github.com/maazkhan/5af0caf3f97b06ee97d57a8d7aa2286d
application.yml - https://gist.github.com/maazkhan/06732e6054c5110c1201afdccfc879ca

**Comment by maazkhan** [comment]
Update: I read the docs and realized that authorization is part of sensitive headers 
http://cloud.spring.io/spring-cloud-static/spring-cloud.html#_cookies_and_sensitive_headers

So now I modified the application.yml to not include authorization as sensitive header and still I don't see authorization header when printing headers in above mentioned Zuul Filter

changes to application.yml
```
zuul:
  routes:
    resource:
      path: /test/**
      sensitiveHeaders: Cookie,Set-Cookie
      url: http://google.com
```

**Comment by dsyer** [comment]
Token relay requires a cookie (to exchange for the token), but I think it works without the sensitive header settings. Your filter is a bit odd though, because it is printing the headers from the incoming request (which Zuul has no control over). I wouldn't be surprised to see no authorization header there if the request came from a browser, for instance.

**Comment by maazkhan** [comment]
Thanks @dsyer ,Yes the request is coming from the browser. I can see the authorization header leaving the http wire in debug logs. How can one grab the authorization header in Zuul filter ?

**Comment by dsyer** [comment]
If your browser is sending an Authorization header, then you are only going to get it relayed to the backend by configuring the sensitive headers (so you are not doing a token relay per spring cloud security, just a straight passthrough of the header). Also, if your browser is sending the header and it isn't arriving at the filter, then something else stripped it out. I don't think it was anything in Spring Cloud. Can you reproduce the problem with a minimal sample app?

**Comment by maazkhan** [comment]
Here is the sample app - https://github.com/maazkhan/OAuth2Proxy
What I am trying to accomplish is -
1. SSO with github
2. Use Zuul to add 2 headers ("testing" header which has token in it, "role" header which has authorities/scope in it.)
3. After that the request is passed to backed service (in this case simply google.com) the request should consist of both "testing" and "role" header

I am having problem in step 3. How can I have "testing" and "role" header visible to backend service

In the logs I can see both of these headers leaving the wire but I never see them at the backend service request when I trace it.

Zuul headers
```
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - GET request to http://localhost:8080/test 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Zuul Headers[role, testing] 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - accept, Value - text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - accept-encoding, Value - gzip, deflate, sdch 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - accept-language, Value - en-US,en;q=0.8 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - connection, Value - keep-alive 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - cookie, Value - rocketchatscreenshare=chrome; JSESSIONID=274941D1D2A9E0D1CDB0522A75CD8876; XSRF-TOKEN=8a2f2019-fff7-42c4-9053-8c0a439536a8 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - host, Value - localhost:8080 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - upgrade-insecure-requests, Value - 1 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Header Name - user-agent, Value - Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Cookie Name - rocketchatscreenshare, Value - chrome 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Cookie Name - XSRF-TOKEN, Value - db83d2fa-d40f-40a3-bb7f-8541155d12a3 
13:01:31.877 [http-nio-8080-exec-4] DEBUG c.s.p.o.filters.AddHeaderFilter - Cookie Name - JSESSIONID, Value - 24611DD0EB37855DC5DB75418FD7FDB4 
```

logs from app - 
```
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> GET / HTTP/1.1 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> accept-language: en-US,en;q=0.8 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> upgrade-insecure-requests: 1 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> authorization: bearer 50c5462db1b6c9e0a2800b96ec48e782e7344895 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> role: [ROLE_USER] 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> x-forwarded-host: localhost 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> x-forwarded-proto: http 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> x-forwarded-prefix: /test 
13:01:31.956 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> testing: bearer 50c5462db1b6c9e0a2800b96ec48e782e7344895 
13:01:31.957 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> x-forwarded-port: 8080 
13:01:31.957 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> x-forwarded-for: 0:0:0:0:0:0:0:1 
13:01:31.957 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> Accept-Encoding: gzip 
13:01:31.957 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> Host: google.com 
13:01:31.957 [http-nio-8080-exec-4] DEBUG org.apache.http.headers - http-outgoing-0 >> Connection: Keep-Alive 
```

header trace sent to backend service
```
Request URL:http://localhost:8080/test
Request Method:GET
Status Code:302 Found
Remote Address:[::1]:8080

Response Headers
view source
Cache-Control:no-cache, no-store, max-age=0, must-revalidate
Content-Length:0
Date:Thu, 19 May 2016 20:01:29 GMT
Expires:0
Location:http://localhost:8080/login
Pragma:no-cache
Server:Apache-Coyote/1.1
Set-Cookie:JSESSIONID=EBD0F566C6448F5A8482CC8B409234F5; Path=/; HttpOnly
Set-Cookie:XSRF-TOKEN=db83d2fa-d40f-40a3-bb7f-8541155d12a3; Path=/
X-Content-Type-Options:nosniff
X-Frame-Options:DENY
X-XSS-Protection:1; mode=block

Request Headers
view source
Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-US,en;q=0.8
Connection:keep-alive
Cookie:rocketchatscreenshare=chrome; JSESSIONID=274941D1D2A9E0D1CDB0522A75CD8876; XSRF-TOKEN=8a2f2019-fff7-42c4-9053-8c0a439536a8
Host:localhost:8080
Upgrade-Insecure-Requests:1
User-Agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
```

**Comment by dsyer** [comment]
I'm not really sure what you are showing there (other than that you gave the world your access token, which was probably an error), or what is working and what is not. How did you trace the backend service (which looks like it is google.com)? What would google.com do with your custom headers anyway?

**Comment by maazkhan** [comment]
The access token above is invalid. I replaced it with some random strings.
I used chrome browser to inspect what packets are leaving from the browser (right click -> inspect on browser). This was just a simple test, eventually I will replace google.com with my own backend service that will have RBAC control based on scopes from OAuth.

I understand custom header won't do anything when sent to google.com but I am interested in seeing that the custom headers were actually sent with the request which I don't see when looking at the network traffic.

**Comment by dsyer** [comment]
I dnt think you can see the outgoing requests from the proxy in your browser (how would it know). It would be better to write your own backend and log the requests as they come in.

**Comment by maazkhan** [comment]
@dsyer Not sure what exactly happened here. I do have a backend service which is printing all the incoming request headers. I didn't saw the headers added via Zuul filter. I was running the OAuth Proxy via bootRun within Intellij IDE. Today I just realize that after changing the dependencies multiple times I never really cleaned the environment. So all I did was ./gradle clean (also cleaned maven cache) and re ran OAuthProxy and everything seems to be working fine. Also I changed my spring-cloud-dependency to Brixton.BUILD-SNAPSHOT. 
The proxy works fine with RC1/RC2 also.
Many thanks for your time!
